packages/micropip/src/micropip/_micropip.py

```diff
@@ -445,6 +468,54 @@ async def install(
     await gather(*wheel_promises)
 
 
+def _generate_package_hash(data: BytesIO) -> str:
+    sha256_hash = hashlib.sha256()
+    data.seek(0)
+    while chunk := data.read(4096):
+        sha256_hash.update(chunk)
+    return sha256_hash.hexdigest()
+
+
+def freeze():
+    from copy import deepcopy
+
+    packages = deepcopy(BUILTIN_PACKAGES)
+    for dist in importlib_distributions():
+        name = dist.name
+        version = dist.version
+        url = dist.read_text("PYODIDE_URL")
+        if url is None:
+            continue
+
+        sha256 = dist.read_text("PYODIDE_SHA256")
+        assert sha256
+        imports = (dist.read_text("top_level.txt") or "").split()
```

Comment:
Probably we should discuss the expected usage of `imports` key in packages.json later. It `meta.yaml` spec, we explained that [it is for testing](https://pyodide.org/en/stable/development/meta-yaml.html#test-imports) and we actually use it in [our test](https://github.com/pyodide/pyodide/blob/60b9344b5622958e9ce7b173b1ec48816a8df424/packages/test_packages_common.py#L72). But it is also used in [loadPackagesFromImports](https://github.com/pyodide/pyodide/blob/60b9344b5622958e9ce7b173b1ec48816a8df424/src/js/api.ts#L127), so it is not only for test, which is a little confusing. (no action required for this PR) (c.c @rth)

Reply:
Maybe we should switch to using `top_level.txt` for `loadPackagesFromImports`.